core/blockchain.go

```diff
@@ -167,11 +173,29 @@ func NewBlockChain(chainDb ethdb.Database, config *params.ChainConfig, engine co
 			}
 		}
 	}
+
+	bc.gc = hashtree.NewGarbageCollector(chainDb, []byte(state.DbPrefix), bc.hasDataCallback)
+
+	/*headBlock := bc.currentBlock.NumberU64()
+	if headBlock > 1000 {
+		bc.gc.FullGC(headBlock - 1000)
+	}*/
+
+	bc.gc.BackgroundGC(bc.CurrentBlock, &bc.processing, &bc.procInterrupt, &bc.wg)
+
 	// Take ownership of this particular state
 	go bc.update()
 	return bc, nil
 }
 
+func (bc *BlockChain) hasDataCallback(version uint64) func(position, hash []byte) bool {
+	header := bc.GetHeaderByNumber(version)
```

Comment:
While I don't yet understand the full context of how this method is used, I do find it a bit odd that `number` is used to resolve a header, since a `number` can be ambiguous. 

So what I'm wondering is if whatever uses this method, does it handle reorgs without breaking?

Reply:
Definitely not. This function verifies a piece of data (identified by hash) being present at the given position in a certain block. This block is the "GC block" which is the oldest block whose state we still want to remember. This block is also the earliest one where we can roll back to. If a longer reorg happens, we should resync the entire chain from the beginning (or do a fast sync). Handling this corner case is not implemented yet.